File: Engine/MonoScripting/Api/EntityComponents/CameraComponent.cpp
#include "../ApiBindings.hpp"
#include <Scene/Camera.hpp>
#include <MonoScripting/MonoScriptingLayer.hpp>

using namespace VSGE;

static float GetNearPlane(Camera* cam) {
	return cam->GetNearPlane();
}
static void SetNearPlane(Camera* cam, float near) {
	cam->SetNearPlane(near);
}

static float GetFarPlane(Camera* cam) {
	return cam->GetFarPlane();
}
static void SetFarPlane(Camera* cam, float far) {
	cam->SetFarPlane(far);
}

static float GetFov(Camera* cam) {
	return cam->GetFOV();
}
static void SetFov(Camera* cam, float fov) {
	cam->SetFOV(fov);
}

static float GetAspectRatio(Camera* cam) {
	return cam->GetAspectRatio();
}
static void SetAspectRatio(Camera* cam, float aspect) {
	cam->SetAspectRatio(aspect);
}

static ViewMask GetViewMask(Camera* cam) {
	return cam->GetViewMask();
}
static void SetViewMask(Camera* cam, ViewMask viewmask) {
	cam->SetViewMask(viewmask);
}

void VSGE::BindCamera() {
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_GetNearPlane(ulong)", GetNearPlane);
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_SetNearPlane(ulong,float)", SetNearPlane);

	MonoScriptingLayer::AddInternalCall("CameraComponent::i_GetFarPlane(ulong)", GetFarPlane);
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_SetFarPlane(ulong,float)", SetFarPlane);

	MonoScriptingLayer::AddInternalCall("CameraComponent::i_GetFov(ulong)", GetFov);
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_SetFarPlane(ulong,float)", SetFov);

	MonoScriptingLayer::AddInternalCall("CameraComponent::i_GetAspectRatio(ulong)", GetAspectRatio);
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_SetAspectRatio(ulong,float)", SetAspectRatio);

	MonoScriptingLayer::AddInternalCall("CameraComponent::i_GetViewMask(ulong)", GetViewMask);
	MonoScriptingLayer::AddInternalCall("CameraComponent::i_SetViewMask(ulong,ulong)", SetViewMask);
}

File: MonoScripting/Api/EngineClass/EntityComponents/CameraComponent.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public class CameraComponent : IEntityComponent {
    public CameraComponent(ulong handle) : base(handle)
    {
    }

    public float GetNearPlane(){
        return i_GetNearPlane(handle);
    }
    public void SetNearPlane(float near){
        i_SetNearPlane(handle, near);
    }

    public float GetFarPlane(){
        return i_GetFarPlane(handle);
    }
    public void SetFarPlane(float near){
        i_SetFarPlane(handle, near);
    }

    public float GetFov(){
        return i_GetFov(handle);
    }
    public void SetFov(float fov){
        i_SetFov(handle, fov);
    }

    public float GetAspectRatio(){
        return i_GetAspectRatio(handle);
    }
    public void SetAspectRatio(float aspect){
        i_SetAspectRatio(handle, aspect);
    }

    public ulong GetViewMask(){
        return i_GetViewMask(handle);
    }
    public void SetViewMask(ulong viewmask){
        i_SetViewMask(handle, viewmask);
    }

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetNearPlane(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetNearPlane(ulong handle, float near);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetFarPlane(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetFarPlane(ulong handle, float far);
   
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetFov(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetFov(ulong handle, float fov);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern float i_GetAspectRatio(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetAspectRatio(ulong handle, float aspect);

    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern ulong i_GetViewMask(ulong handle);
    [MethodImpl(MethodImplOptions.InternalCall)]
    private static extern void i_SetViewMask(ulong handle, ulong viewmask);
}